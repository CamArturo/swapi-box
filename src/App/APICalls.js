export const fetchVehicles = vehicles => {
  try {
    const results = vehicles.map(vehicle => {
      const name = vehicle.name;
      const model = vehicle.model;
      const vehicleClass = vehicle.vehicle_class;
      const passengers = vehicle.passengers;
      return {
        name: name,
        model: model,
        vehicleClass: vehicleClass,
        passengers: passengers,
        favorite: false
      };
    });
    return Promise.all(results);
  } catch (error) {
    this.setState({
      errorStatus: 'Error adding Vehicle'
    });
  }
};


export const fetchCategory = async (category) => {
  try {
    const url = `https://swapi.co/api/${category}/`;
    const response = await fetch(url);
    const categoryUrls = await response.json();
    this.setState({
      [`${category}Urls`]: categoryUrls.results
    });
  } catch (error) {
    this.setState({
      errorStatus: 'Error adding Category'
    });
  }
};