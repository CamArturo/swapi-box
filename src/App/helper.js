export const getRandomInt = ((min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min;
});

export const isAlreadyFavorite = ((favorites, cardInfo) => {
  return favorites.includes(cardInfo);
});

export const selectCards = ((category, people, planets, vehicles, favorites) => {
  let selectedCards = [];
  if (category === 'people') {
    selectedCards = people;
  } else if (category === 'planets') {
    selectedCards = planets;
  } else if (category === 'vehicles') {
    selectedCards = vehicles;
  } else {
    selectedCards = favorites;
  }
  return selectedCards;
});