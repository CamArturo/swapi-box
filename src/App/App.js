import React, {Component} from 'react';
import Scroller from './Components/StatelessComponents/Scroll/Scroller';
import Header from './Components/StatelessComponents/Header/Header';
import Navigation from './Components/StatelessComponents/Navigation/Navigation';
import CardContainer
  from "./Components/StatelessComponents/CardContainer/CardContainer";
import Loading from "./Components/StatelessComponents/Loading/Loading";
import {fetchVehicles, fetchCategory} from './APICalls';
import {getRandomInt, isAlreadyFavorite} from './helper';
import './App.css';


//Components
// APICalls
// container with redux

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      scroller: true,
      filmsInfo: {},
      data: [],
      people: [],
      vehicles: [],
      planets: [],
      favorites: [],
      currentCategory: ''
    };
  }

  updateFavorites = ((cardInfo, category) => {
    const isFound = isAlreadyFavorite(this.state.favorites, cardInfo);
    if (!isFound) {
      cardInfo.favorite = true;
      this.setState({
        favorites: [...this.state.favorites, cardInfo]
      });
    } else {
      let arrayData = this.state[`${category}`];
      let modifiedArray = arrayData.map((object) => {
        if (object.name === cardInfo.name) {
          cardInfo.favorite = !cardInfo.favorite;
        }
        return object;
      });
      const filteredModifiedArray = modifiedArray.filter(item => {
        return item.favorite === true;
      });
      this.setState({
        favorites: filteredModifiedArray
      });
    }
  });

  displayFavorites = () => {
    this.setState({
      currentCategory: 'favorites',
      loading: false
    });
  };

  fetchPerson = people => {
    try {
      const results = people.map(async person => {
        const name = person.name;
        const responseHomeWorldUrl = person.homeworld;
        // TODO make following 2 lines a function for testing
        const responseHomeWorld = await fetch(responseHomeWorldUrl);
        // TODO 200 response
        // if (responseHomeWorld.status < 300) {
        //
        // }
        const homeWorldData = await responseHomeWorld.json();
        const speciesUrl = person.species;
        const responseSpecies = await fetch(speciesUrl);
        const speciesData = await responseSpecies.json();
        return {
          name: name,
          species: speciesData.name,
          homeworld: homeWorldData.name,
          population: homeWorldData.population,
          favorite: false
        };
      });
      return Promise.all(results);
    } catch (error) {
      this.setState({
        errorStatus: 'Error adding Person'
      });
    }
  };

  fetchPlanets = planets => {
    try {
      const results = planets.map(async planet => {
        const name = planet.name;
        const terrain = planet.terrain;
        const population = planet.population;
        const climate = planet.climate;
        const residentsUrl = planet.residents;
        const residentsData = await this.fetchResidents(residentsUrl);
        const residentNames = residentsData.map(resident => {
          return ' ' + resident.name;
        });
        const residentString = residentNames.toString();
        return {
          name: name,
          terrain: terrain,
          population: population,
          climate: climate,
          residents: residentString,
          favorite: false
        };
      });
      return Promise.all(results);
    } catch (error) {
      this.setState({
        errorStatus: 'Error adding Planet'
      });
    }
  };

  fetchResidents = residentsUrl => {
    try {
      const results = residentsUrl.map(async residentUrl => {
        if (residentUrl.length > 1) {
          const responseResidents = await fetch(residentUrl);
          return await responseResidents.json();
        }
        return results;
      });
      return Promise.all(results);
    } catch (error) {
      this.setState({
        errorStatus: 'Error adding Planet'
      });
    }
  };

  fetchCategory = async (category) => {
    try {
      const url = `https://swapi.co/api/${category}/`;
      const response = await fetch(url);
      const categoryUrls = await response.json();

      // const urls = fetchCategory(category)

      this.setState({
        [`${category}Urls`]: categoryUrls.results
      });
    } catch (error) {
      this.setState({
        errorStatus: 'Error adding Category'
      });
    }
  };

  updateCards = async (category) => {
    this.setState({
      scroller: false
    });
    switch (category) {
      case 'people':
        let people;
        if (!this.state.people.length) {
          people = await this.fetchPerson(this.state.peopleUrls);
        } else {
          people = this.state.people;
        }
        this.setState({
          people,
          currentCategory: 'people',
          loading: false
        });
        break;
      case 'planets':
        let planets;
        if (!this.state.planets.length) {
          planets = await this.fetchPlanets(this.state.planetsUrls);
        } else {
          planets = this.state.planets;
        }
        this.setState({
          planets,
          currentCategory: 'planets',
          loading: false
        });
        break;
      case 'vehicles':
        let vehicles;
        if (!this.state.vehicles.length) {
          vehicles = await fetchVehicles(this.state.vehiclesUrls);
        } else {
          vehicles = this.state.vehicles;
        }
        this.setState({
          vehicles,
          currentCategory: 'vehicles',
          loading: false
        });
        break;
      default:
        break;
    }
  };

  async componentDidMount() {
    const randomNum = getRandomInt(1, 7);
    const url = `https://swapi.co/api/films/${randomNum}`;
    const response = await fetch(url);
    const data = await response.json();
    const filmsInfo = {
      title: data.title,
      date: data.release_date,
      scroll: data.opening_crawl
    };
    this.setState({filmsInfo});
    await this.fetchCategory('people');
    await this.fetchCategory('planets');
    await this.fetchCategory('vehicles');
  }

  render() {
    return (
      <div className="App">
        <Header/>
        {
          this.state.filmsInfo &&
          this.state.scroller &&
          <Scroller
            filmsInfo={this.state.filmsInfo}
          />
        }
        {
          !this.state.scroller &&
          this.state.loading &&
          <Loading/>
        }
        <CardContainer
          category={this.state.currentCategory}
          people={this.state.people}
          planets={this.state.planets}
          vehicles={this.state.vehicles}
          favorites={this.state.favorites}
          updateFavorites={this.updateFavorites}
        />
        <Navigation
          currentCategory={this.state.currentCategory}
          selectCategory={this.updateCards}
          displayFavorites={this.displayFavorites}
          favorites={this.state.favorites.length}
        />
      </div>
    );
  }
}

export default App;