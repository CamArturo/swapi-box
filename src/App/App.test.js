import React from 'react';
import {shallow} from 'enzyme';
import App from './App';


describe('App test', () => {
  let wrapper;
  let category;
  let mockData;

  beforeEach(() => {
    // Mock fetch - need to able to do normal stuff with it.
    category = 'vehicles';
    mockData = {
      "results": [
        {
          "name": "Sand Crawler",
          "model": "Digger Crawler",
          "manufacturer": "Corellia Mining Corporation",
          "cost_in_credits": "150000",
          "length": "36.8",
          "max_atmosphering_speed": "30",
          "crew": "46",
          "passengers": "30",
          "cargo_capacity": "50000",
          "consumables": "2 months",
          "vehicle_class": "wheeled",
          "pilots": [],
          "films": [
            "https://swapi.co/api/films/5/",
            "https://swapi.co/api/films/1/"
          ]
        }]
    };

    // Return an obj - key in resolve is JSON - a key in our object
    // after
    // window.fetch = jest.fn().mockImplementation(() => Promise.resolve({json: () => Promise.resolve(mockData)}));
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({json: () => Promise.resolve(mockData), status: 200}));
    // expect whatever you want to do if response .status is not 200
    wrapper = shallow(<App/>);
    // Actions

    //Expectation
  });
  it('renders without crashing', () => {
    const wrapper = shallow(<App/>, {
      disableLifecycleMethods: true
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('should call fetch with the correct args', async () => {
    const expected = 'https://swapi.co/api/vehicles/';

    await wrapper.instance().fetchCategory(category);
    expect(window.fetch).toHaveBeenCalledWith(expected);
  });

  it('should set state with the returned data', async () => {
    const expected = [
      {
        "name": "Sand Crawler",
        "model": "Digger Crawler",
        "manufacturer": "Corellia Mining Corporation",
        "cost_in_credits": "150000",
        "length": "36.8",
        "max_atmosphering_speed": "30",
        "crew": "46",
        "passengers": "30",
        "cargo_capacity": "50000",
        "consumables": "2 months",
        "vehicle_class": "wheeled",
        "pilots": [],
        "films": [
          "https://swapi.co/api/films/5/",
          "https://swapi.co/api/films/1/"
        ]
      }];
    await wrapper.instance().fetchCategory(category);
    expect(wrapper.state('vehiclesUrls')).toEqual(expected);
  });

  it('should update error status in state', async () => {
    // window.fetch mocking fetch function
    window.fetch = jest.fn().mockImplementation(() => Promise.reject(new Error('text')));
    await wrapper.instance().fetchCategory(category);
    expect(wrapper.state('errorStatus')).toEqual('Error adding Category');
  });


  // promise resolves
});
