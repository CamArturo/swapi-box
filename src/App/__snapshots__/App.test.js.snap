// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <AudioPlayer />,
        <Header />,
        <Scroller
          filmsInfo={Object {}}
/>,
        <CardContainer
          category=""
          favorites={Array []}
          people={Array []}
          planets={Array []}
          updateFavorites={[Function]}
          vehicles={Array []}
/>,
        <Navigation
          displayFavorites={[Function]}
          favorites={0}
          selectCategory={[Function]}
/>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "filmsInfo": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "category": "",
          "favorites": Array [],
          "people": Array [],
          "planets": Array [],
          "updateFavorites": [Function],
          "vehicles": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "displayFavorites": [Function],
          "favorites": 0,
          "selectCategory": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <AudioPlayer />,
          <Header />,
          <Scroller
            filmsInfo={Object {}}
/>,
          <CardContainer
            category=""
            favorites={Array []}
            people={Array []}
            planets={Array []}
            updateFavorites={[Function]}
            vehicles={Array []}
/>,
          <Navigation
            displayFavorites={[Function]}
            favorites={0}
            selectCategory={[Function]}
/>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "filmsInfo": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "category": "",
            "favorites": Array [],
            "people": Array [],
            "planets": Array [],
            "updateFavorites": [Function],
            "vehicles": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "displayFavorites": [Function],
            "favorites": 0,
            "selectCategory": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
